### PR TITLE
ADT for JSON schema, pretty print interpreters & encoder, decoder pairs

### PR DESCRIPTION
**Pull request by lookageek**
This PR constructs the JSON ADT to capture the JSON schema in its entirety.

It provides these interpreters for JSON ADT to be operated on -

- `prettyPrint` - converts the JSON ADT object into a string using string representation rules of JSON schema
- `removeNull` - removes key-value pairs from JSON data where the value is of JSON type Null
- `skipNullPrettyPrint` - an interpreter made from composing removeNull and prettyPrint functions together - creating string representation of JSON removed all key-value pairs whose values are JSON Null

Also, PR includes the encoder decoder pair for these JSON types

`String`, `Int`, `Boolean`, `List`, `Vector`, `Option`, `Either`, Tuples 2 & 3, and `User` data type

### CODE REVIEW

File: json/src/main/scala/com/jpmc/json/Json.scala [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+  final case class String(value: Predef.String) extends Json
+
+  final case class Int(value: scala.Int) extends Json
+
+  final case class Double(value: scala.Double) extends Json
+
+  final case class Bool(value: scala.Boolean) extends Json
+
+  sealed trait Doc extends Json
```
Comment (senertopaloglu) [inline json/src/main/scala/com/jpmc/json/Json.scala:16]: It probably skipped your mind while coding, but I think it’s a good practice to use `sealed trait Doc extends Json with Product with Serializable` here. It stops namespace pollution and Nicolas had an example in one of his talks, I’ll paste a link here if I find it.

File: codecs/src/main/scala/com/jpmc/codecs/Decoder.scala [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  }.withError("list decode error")
+
+  implicit def vectorDecoder[A: Decoder]: Decoder[Vector[A]] = listDecoder[A].map(_.toVector)
+
+  implicit def tuple2Decoder[A: Decoder, B: Decoder]: Decoder[(A, B)] = fromPartial {
+    case Json.Array(List(a, b)) => for {
+      aRes <- a.as[A]
+      bRes <- b.as[B]
+    } yield (aRes, bRes)
+  }.withError("tuple2 decode error")
```
Comment (mram411) [inline codecs/src/main/scala/com/jpmc/codecs/Decoder.scala:98]: I like this tuple encoder/decoder implementation.

File: codecs/src/main/scala/com/jpmc/codecs/Encoder.scala [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+  def contraMap[B](f: B => A): Encoder[B] = Encoder.from { b => this.encode(f(b)) }
+}
+
+object Encoder {
+  def apply[A](implicit encoder: Encoder[A]): Encoder[A] = encoder
+
+  def from[A](f: A => Json): Encoder[A] = (value: A) => f(value)
+
+  implicit class EncoderSyntax[A](value: A) {
```
Comment (monmcguigan) [inline codecs/src/main/scala/com/jpmc/codecs/Encoder.scala:19]: Why do you have to define the asJson function in this class like this?

Comment (lookageek) [reply codecs/src/main/scala/com/jpmc/codecs/Encoder.scala:19]: thanks for the review @monmcguigan 

This pattern is known as "extension methods" or "pimp my class" (analogy to MTVs "pimp my ride" show)
It enables one to define a method on a type, and use it as if the type had the method already.
Consider the `Int` class, one can define a method to be used on integers! 
```scala

// Int is the type to which new methods are being added
implicit class PimpMyInteger(value: Int) {    
   def beefUp(by: Int) {
      value * by
   }
   
   // can add any number of methods
   def toneDown(by: Int) {
     value / by 
   }
}
```

now this new method can be used as a nifty way to multiple two numbers or divide them

```scala

val beefedUp = 2.beefUp(3)
// OR
val beefedUp = 2 beefUp 3

// take it down a notch
val tonedDown = 100 toneDown 4
```
`beefUp` is something which scala language does not provide give on the integer type 😆, but we made it look like this method came from integer type in scala language.

In scala 2 this involves creating an `implicit class`
This pattern has been syntactically formalised in [scala 3 with `extension` keyword](https://docs.scala-lang.org/scala3/reference/contextual/extension-methods.html)
Another language which has this feature is [Kotlin](https://kotlinlang.org/docs/extensions.html) 

With the `EncoderSyntax` class that is defined here, it is in an abstraction level above - the type on which the extension method is defined itself is a parameter `EncoderSyntax[A]`, and `asJson` will work as long as the type on which `asJson` is being called has an implicit instance of the Encoder in the scope!

Comment (monmcguigan) [reply codecs/src/main/scala/com/jpmc/codecs/Encoder.scala:19]: ahhhhh cool!! Thanks for this explanation, that makes complete sense.

**Pull request merged**